go/vt/wrangler/reparent.go

```diff
@@ -996,55 +995,279 @@ func (wr *Wrangler) emergencyReparentShardLocked(ctx context.Context, ev *events
-					forceStart = replicaWasRunning(status)
-				}
-				if err := wr.tmc.SetMaster(replCtx, tabletInfo.Tablet, masterElectTabletAlias, now, "", forceStart); err != nil {
-					rec.RecordError(fmt.Errorf("tablet %v SetMaster failed: %v", alias, err))
-				}
-			}(alias, tabletInfo)
+			go handleMaster(alias, tabletInfo)
+		} else if !unreachableReplicas.Has(alias) {
+			go handleReplicas(alias, tabletInfo)
 		}
 	}
 
 	wgMaster.Wait()
 	if masterErr != nil {
-		// The master failed, there is no way the
-		// replicas will work.  So we cancel them all.
-		wr.logger.Warningf("master failed to PopulateReparentJournal, canceling replicas")
+		wr.logger.Warningf("master failed to PopulateReparentJournal")
 		replCancel()
-		wgReplicas.Wait()
+		waitOnTablets()
 		return fmt.Errorf("failed to PopulateReparentJournal on master: %v", masterErr)
 	}
 
-	// Wait for the replicas to complete. If some of them fail, we
-	// will rebuild the shard serving graph anyway
-	wgReplicas.Wait()
-	if err := rec.Error(); err != nil {
-		wr.Logger().Errorf2(err, "some replicas failed to reparent")
+	errorRecorder := waitOnTablets()
+	if len(errorRecorder.Errors) > 1 {
+		// We allow one error, for the presumed non-contactable dead master tablet.
+		wr.Logger().Errorf2(errorRecorder.Error(), "some replicas failed to reparent")
 		return err
 	}
 
 	return nil
 }
 
+// waitOnNMinusOneTablets will wait until N-1 tablets have responded via a supplied error channel. In that case that N-1 tablets have responded,
+// the supplied cancel function will be called, and we will wait until N tablets return their errors, and then return an AllErrorRecorder to the caller.
+func waitOnNMinusOneTablets(ctxCancel context.CancelFunc, tabletCount int, errorChannel chan error) *concurrency.AllErrorRecorder {
+	errCounter := 0
+	successCounter := 0
+	responseCounter := 0
+	rec := &concurrency.AllErrorRecorder{}
+
+	for err := range errorChannel {
+		responseCounter++
+		successCounter++
+		if err != nil {
+			errCounter++
+			successCounter--
+			rec.RecordError(err)
+		}
+		if responseCounter == tabletCount {
+			// We must wait for any cancelled goroutines to return their error.
+			break
+		}
+		if errCounter > 1 || successCounter == tabletCount-1 {
+			ctxCancel()
+		}
+	}
+
+	return rec
+}
+
+// findValidReparentCandidates will find valid candidates for emergency reparent, and if successful, returning them as a list of tablet aliases.
+func (wr *Wrangler) findValidReparentCandidates(statusMap map[string]*replicationdatapb.StopReplicationStatus, masterStatusMap map[string]*replicationdatapb.MasterStatus) ([]string, error) {
+	// Build out replication status list from proto types.
+	replicationStatusMap := make(map[string]*mysql.ReplicationStatus, len(statusMap))
+	for alias, protoStatus := range statusMap {
+		status := mysql.ProtoToReplicationStatus(protoStatus.After)
+		replicationStatusMap[alias] = &status
+	}
+
+	// Determine if we need to find errant GTIDs.
+	var gtidBased *bool
```

Comment:
hmmm, I'm not sure why I thought I needed to treat the first case as different (know if it had been set yet, which is why I made this a pointer bool). I think you're right though, I'll change it.

Reply:
Oh, I'm seeing why now. Because with the approach where we assume true, we can't bail if we find an odd one in the bunch. In other words, unless we know for sure we actually started on true, then we can't error out if we find one that has a relay log position missing.